src/detect-engine-alert.c

```diff
@@ -338,71 +336,136 @@ static inline void FlowApplySignatureActions(
+        uint16_t tmp_size = 0;
+        bool is_action_pass = false;
+        uint32_t pass_sid_num = 0;
+
+        while (i < max_pos) {
+            const Signature *s = de_ctx->sig_array[det_ctx->alert_queue[i].num];
+            uint8_t res = PacketAlertFinalize(de_ctx, det_ctx, p, &det_ctx->alert_queue[i], s);
+
+            /* Thresholding removes this alert */
+            if (res == 0) {
+                /* we will not copy this to the AlertQueue */
+                p->alerts.suppressed++;
+                SCLogDebug("Suppressing sid %" PRIu32 " alert from alerts' queue", s->id);
+            } else if (res == 1) {
+                tmp_queue[tmp_size] = det_ctx->alert_queue[i];
+                tmp_size++;
+                SCLogDebug("Appending sid %" PRIu32 " alert to Packet::alerts at pos %u", s->id, i);
+                if (PacketTestAction(p, ACTION_PASS) && !is_action_pass) {
+                    /* Ok, reset the alert cnt to end in the previous of pass
+                     * so we ignore the rest with less prio */
+                    pass_sid_num = s->num;
+                    is_action_pass = true;
                 }
             }
+            i++;
+        }
 
-            /* set actions on the flow */
-            FlowApplySignatureActions(
-                    p, &det_ctx->alert_queue[i], s, det_ctx->alert_queue[i].flags);
+        if (packet_alert_max > tmp_size) {
+            p->alerts.cnt = tmp_size;
+        } else {
+            p->alerts.cnt = packet_alert_max;
+            // we also have discarded alerts in case of queue expansion failure, let's
+            // not miss that count
+            p->alerts.discarded += (tmp_size - packet_alert_max);
+        }
+        qsort(tmp_queue, tmp_size, sizeof(PacketAlert), AlertQueueSortHelper);
 
-            /* set actions on packet */
-            PacketApplySignatureActions(p, s, det_ctx->alert_queue[i].flags);
+        if (det_ctx->is_alert_queue_expand_failure) {
+            if (p->alerts.drop.action & ACTION_DROP) {
+                PacketAlertFinalize(de_ctx, det_ctx, p, &p->alerts.drop, p->alerts.drop.s);
+            }
         }
 
-        /* Thresholding removes this alert */
-        if (res == 0 || res == 2 || (s->flags & SIG_FLAG_NOALERT)) {
-            /* we will not copy this to the AlertQueue */
-            p->alerts.suppressed++;
-        } else if (p->alerts.cnt < packet_alert_max) {
-            p->alerts.alerts[p->alerts.cnt] = det_ctx->alert_queue[i];
-            SCLogDebug("Appending sid %" PRIu32 " alert to Packet::alerts at pos %u", s->id, i);
-
-            if (PacketTestAction(p, ACTION_PASS)) {
-                /* Ok, reset the alert cnt to end in the previous of pass
-                 * so we ignore the rest with less prio */
-                break;
+        if (is_action_pass) {
+            // this means we got a 'PASS' action for this Packet, pass_sid_num
+            // has the related signature's internal id
+            uint16_t j;
```

Comment:
would prefer j to be initialized to 0 here and not in the for statement, as its used outside of the for loop as well